### ISSUE TITLE
@types/jest V24.0.20 produces error: TS2428: All declarations of 'Matchers' must have identical type parameters

### ISSUE THREAD

**Issue by iain-davis** [DefinitelyTyped/DefinitelyTyped#39982]
If you know how to fix the issue, make a pull request instead.

- [X] I tried using the `@types/jest` package and had problems.
- [X] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [X] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [X] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @NoHomey @jwbay @asvetliakov @alexjoverm @epicallan @ikatyang @wsmd @JamieMason @douglasduteil @ahnpnl @JoshuaKGoldberg @UselessPickles @r3nya @Hotell @sebald @andys8 @antoinebrault @Favna @gstamac @ExE-Boss @quassnoi @Belco90 @tonyhallett 

After updating @types/jest to 24.0.20, we had the following error. We were able to fix the issue by downgrading to 24.0.19, so this appears to be an issue with the changes in 24.0.20.

`｢atl｣: Using typescript@3.6.4 from typescript                                                                                              
｢atl｣: Using tsconfig.json from C:/Projects/projectname/tsconfig.json                                                               
｢atl｣: Checking started in a separate process...                                                                                           
｢atl｣: Checking finished with 2 errors                                                                                                     
Child                                                                                                                                        
                                                                                                                                           
  ERROR in [at-loader] ./node_modules/jest-enzyme/lib/index.d.ts:7:15                                                                      
      TS2428: All declarations of 'Matchers' must have identical type parameters.                                                          
                                                                                                                                           
  ERROR in [at-loader] ./node_modules/@types/jest/index.d.ts:639:15                                                                        
      TS2428: All declarations of 'Matchers' must have identical type parameters.                                                          
pm ERR! code ELIFECYCLE                                                                                  `

**Comment by ExE-Boss** [comment]
Yeah, that’s because of #39243.

**Comment by iain-davis** [comment]
I read through #39243, but I didn't discern from that issue what steps I should be taking. Do I need to wait until for a change to @types/jest? Wait for jest-enzyme to catch up to the change in @types/jest? Or is there another package I can update to resolve my error? For now, I have pinned the version at 24.0.19.

**Comment by ExE-Boss** [comment]
You need to wait for `jest-enzyme` to add the new `T` type parameter to `Matchers`.

**Comment by kbradl16** [comment]
I see this in the types file
`// should be R extends void|Promise<void> but getting dtslint error
    interface Matchers<R, T = {}> {`

jest-extended has this `interface Matchers<R, T = void> { `
https://github.com/jest-community/jest-extended/blob/master/types/index.d.ts#L5

Which is right? should it be `T = {}` or `T = void` ?

**Comment by tonyhallett** [comment]
T = {}
See [comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41766#discussion_r369369310)

R should be either void or Promise<void> but definitely typed files go through dtslint which was erroneously preventing this.

**Comment by tonyhallett** [comment]
@kbradl16 
On further thoughts I think T should just be T.

I also think that there is an error with [toContainEntry](https://github.com/jest-community/jest-extended/blob/b10c81c3d880f5d05b1c362e098a607343e4b83e/types/index.d.ts#L255).  Also probably with toContainEntries, toContainAllEntries and toContainAnyEntries.

```
toContainEntry<T>(entry: [keyof T, T[keyof T]]): R;
```

I think that it should be toContainEntry(entry: [keyof T, T[keyof T]]): R;
( The [Expect interface](https://github.com/jest-community/jest-extended/blob/b10c81c3d880f5d05b1c362e098a607343e4b83e/types/index.d.ts#L611) on the other hand is correct. )

This is the [source code](https://github.com/jest-community/jest-extended/blob/master/src/matchers/toContainEntry/index.js)

Given that testing property of T (TActual) - key should come from that T, not some other T.
```
export default {
  toContainEntry: (actual, expected) => {
    const pass = predicate(actual, expected);
    /*
        predicate code
        export default (obj, [key, value]) => obj.hasOwnProperty && obj.hasOwnProperty(key) && equals(obj[key], value);
    */
    if (pass) {
      return { pass: true, message: passMessage(actual, expected) };
    }

    return { pass: false, message: failMessage(actual, expected) };
  }
};

```
predicate.js
```
export default (obj, [key, value]) => obj.hasOwnProperty && obj.hasOwnProperty(key) && equals(obj[key], value);
```

@mattphillips 
How does #41766 make a difference ?

With respect to declaration merging, if I do the following

```
declare global {
  namespace jest {
  
  interface Matchers<R>{}
  
  }
}
```
I get error message - All declarations of 'Matchers' must have identical type parameters.

Under what circumstances is it benefical ?

----

Also given that Matchers are used for the return type of expect and expect is typed as :

```
<T = any>(actual: T): JestMatchers<T>;
type JestMatchers<T> = JestMatchersShape<Matchers<void, T>, Matchers<Promise<void>, T>>;
```
T should default to type any rather than {}.

**Comment by tonyhallett** [comment]
@kbradl16 
On further thoughts I think T should just be T.  The default was there for backwards compatibility, given that it is present T should be T.

I also think that there is an error with [toContainEntry](https://github.com/jest-community/jest-extended/blob/b10c81c3d880f5d05b1c362e098a607343e4b83e/types/index.d.ts#L255).  Also probably with toContainEntries, toContainAllEntries and toContainAnyEntries.

```
toContainEntry<T>(entry: [keyof T, T[keyof T]]): R;
```

I think that it should be toContainEntry(entry: [keyof T, T[keyof T]]): R;
( The [Expect interface](https://github.com/jest-community/jest-extended/blob/b10c81c3d880f5d05b1c362e098a607343e4b83e/types/index.d.ts#L611) on the other hand is correct. )

This is the [source code](https://github.com/jest-community/jest-extended/blob/master/src/matchers/toContainEntry/index.js)

Given that testing property of T (TActual) - key should come from that T, not some other T.
```
export default {
  toContainEntry: (actual, expected) => {
    const pass = predicate(actual, expected);
    /*
        predicate code
        export default (obj, [key, value]) => obj.hasOwnProperty && obj.hasOwnProperty(key) && equals(obj[key], value);
    */
    if (pass) {
      return { pass: true, message: passMessage(actual, expected) };
    }

    return { pass: false, message: failMessage(actual, expected) };
  }
};

```
predicate.js
```
export default (obj, [key, value]) => obj.hasOwnProperty && obj.hasOwnProperty(key) && equals(obj[key], value);
```

@mattphillips 
Given that Matchers are used for the return type of expect and expect is typed as :

```
<T = any>(actual: T): JestMatchers<T>;
type JestMatchers<T> = JestMatchersShape<Matchers<void, T>, Matchers<Promise<void>, T>>;
```
T should default to type any rather than {}.

**Comment by ImranK1506** [comment]
I included "@types/jest": "26.0.19". Works on my machine.

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
